src/box/applier.cc

```diff
@@ -1496,6 +1518,8 @@ applier_process_batch(struct cmsg *base)
 			stailq_first_entry(&tx->rows, struct applier_tx_row,
 					    next);
 		raft_process_heartbeat(box_raft(), applier->instance_id);
+		if (applier_process_sync(applier, txr->row.sync) != 0)
+			diag_raise();
```

Comment:
We should. I process sync twice for simplicity. First row sync for all messages(like raft, heartbeat and so on) and then last row for transactions:
https://github.com/tarantool/tarantool/pull/7611/files#diff-a99e106ecbdb0944db62f223fa59b3140ea961842ef5c2c139f0a603e1989d9fL1266-L1267

Reply:
Why the last row wouldn't be enough though?

Reply:
It would. I just didn't want to fetch the last row one more time. (It's done already in applier_apply_tx).
But I guess this doesn't matter much, so changed:
<details><summary>diff</summary>

```diff
diff --git a/src/box/applier.cc b/src/box/applier.cc
index 28ad0a430..f0d782e98 100644
--- a/src/box/applier.cc
+++ b/src/box/applier.cc
@@ -1285,8 +1285,6 @@ applier_apply_tx(struct applier *applier, struct stailq *rows)
        struct xrow_header *first_row = &txr->row;
        struct xrow_header *last_row;
        last_row = &stailq_last_entry(rows, struct applier_tx_row, next)->row;
-       if (applier_process_sync(applier, last_row->sync) != 0)
-               return -1;
        struct replica *replica = replica_by_id(first_row->replica_id);
        int rc = 0;
        /*
@@ -1517,14 +1515,14 @@ applier_process_batch(struct cmsg *base)
        struct applier *applier = msg->base.applier;
        struct applier_tx *tx;
        stailq_foreach_entry(tx, &msg->txs, next) {
-               struct applier_tx_row *txr =
-                       stailq_first_entry(&tx->rows, struct applier_tx_row,
-                                           next);
+               struct applier_tx_row *last_txr =
+                       stailq_last_entry(&tx->rows, struct applier_tx_row,
+                                         next);
                raft_process_heartbeat(box_raft(), applier->instance_id);
-               if (applier_process_sync(applier, txr->row.sync) != 0)
+               if (applier_process_sync(applier, last_txr->row.sync) != 0)
                        diag_raise();
-               if (txr->row.lsn == 0) {
-                       if (applier_handle_raft(applier, txr) != 0)
+               if (last_txr->row.lsn == 0) {
+                       if (applier_handle_raft(applier, last_txr) != 0)
                                diag_raise();
                        applier_signal_ack(applier);
                        applier_check_sync(applier);
```

</details>